elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,248 @@
+
+A similar approach will also be taken for the management operations against the credential store, presently the resource contains an `add-alias` operation which can be used to add a credential based on a clear text String.  The following operations will be added to the `credential-store` resource: -
+
+ * `generate-secret-key`
+ * `export-secret-key`
+ * `import-secret-key`
+ 
+Within the management model operations are self describing so making use of dedicated operations makes it easier for tooling to present a meaningful UI to end users automatically.
+
+As with the command line utility it is expected each of these will need to take the alias as a parameter.
+
+NOTE: The credential store also offers a programatic API, should the tooling and operations be insufficient for an end user manual population of the credential store may remain an option.
+
+==== Expression Resolver
+
+Presently WildFly Core supports an expression resolver that can delegate to a Vault configuration and if that is not available fall back to use either system properties or environment variables.  This will be updated to make use of the `CapabilityRegistry` and attempt to lookup a capability using a pre-defined constant (To be defined) which exposes an expression resolver runtime API.  The expression resolver looked up using a capability will be used first to attempt to resolve any expression, if this is successful the resolved value will be used - if not we will fallback to the existing behaviour of using the vault and if that fails using a system property or environment variable.
+
+A new singleton resource will be added to the elytron subsystem called `expression=encryption` the purpose of this resource is to contain the configuration both for the encryption and decryption of values.  As a singleton we know only one instance of this resource can be defined within the subsystem reducing some of the complexity multiple instances would cause.  The resource will contain the definition of one or more resolvers to decrypt the inline expressions.
+
+The format of the expressions resolved using the WildFly Elytron expression resolver will be `${ENC:Resolver:ENCRYPTED_DATA}` where `Resolver` is a reference to the specific resolver defined in the `expression=encryption` resource and `ENCRYPTED_DATA` is the data to be decrypted encoded using Base64.
+
+The `expression=encryption` resource will register it's expression resolver capability with the `CapabilityRegistry` making it available for runtime resolution of expressions.
+
+The `expression=encryption` resource will contain three attributes, the first attribute being `resolvers` which will be used to define one or more resolvers to handle the encryption and decryption of values.  Each resolver will support the following configuration: -
+
+ * Reference to a credential store and alias to identify a specific previously created `SecretKey` to use.
+ * Name of the algorithm to use for the encryption / decryption.
```

Comment:
No the algorithm configuration can be relaxed so will not be needed, one point I need to confirm is if we want to make the size of the initialisation vector configurable so that may be one that gets added.

Reply:
Actually this will need to be adjusted, the attribute will be `transformation` not algorithm, I have added details for the command line tool that will be duplicated in this section but there will be a set of algorithm name / algorithm mode / padding mode permutations that will need to be specified as the transformation.